### PR TITLE
Disable legalhold for user

### PR DESCRIPTION
**Pull request by ChrisPenner**
DELETE:

- [x] Permissions: Only Team "admin?"
- [x] Remove Legal hold device from user. (thus triggering an event; notify clients they no longer need approval if deleted when still pending)
- [x] Can't DELETE LH clients using regular /clients calls
- [x] Can't ADD LH clients using regular /clients calls

- [x] Check that disallowing LHClient deletion doesn't break user deletion workflows

### CODE REVIEW

File: libs/galley-types/src/Galley/Types/Teams.hs [hunk 1]
```diff
@@ -401,6 +401,7 @@ data HiddenPerm
     | ViewLegalHoldTeamSettings
     | ChangeLegalHoldUserSettings
     | ViewLegalHoldUserSettings
+    | DisableLegalHold
```
Comment (fisx) [inline libs/galley-types/src/Galley/Types/Teams.hs]: My plan was to use `ChangeLegalHoldUserSettings` for that (change == add or remove).  If you want to keep `DisableLegalHold`, you probably also want to rename the others to make it consistent again.

File: services/brig/src/Brig/API/Client.hs [hunk 2]
```diff
@@ -163,3 +169,10 @@ legalHoldClientRequested (LegalHoldClientRequest requester targetUser lastPrekey
     eventData = LegalHoldClientRequestedData requester targetUser lastPrekey' clientId
     lhClientEvent :: ClientEvent
     lhClientEvent = LegalHoldClientRequested eventData
+
+removeLegalHoldClient :: UserId -> AppIO ()
+removeLegalHoldClient uid = do
+    clients <- Data.lookupClients uid
+    -- Should only be one; but just in case we'll treat it as a list
+    let legalHoldClients = filter ((== LegalHoldClientType) . clientType) clients
+    forM_ legalHoldClients  (execDelete uid Nothing)
```
Comment (fisx) [inline services/brig/src/Brig/API/Client.hs]: I don't know if there should be a warning in the logs if there is ever more than one LH client?  Or even an error?

File: services/brig/test/resources/turn/servers-v2.txt [hunk 3]
```diff
@@ -1,2 +1 @@
-turn:localhost:3478
-turn:localhost:3479
+turn:localhost:3478
```
Comment (fisx) [inline services/brig/test/resources/turn/servers-v2.txt]: what is this about?  can't find anything in the commit messages.

File: services/galley/src/Galley/Intra/Client.hs [hunk 4]
```diff
@@ -84,17 +85,28 @@ addLegalHoldClientToUser uid connId prekeys lastPrekey' = do
...
+        . expect2xx
+
 brigAddClient :: UserId -> ConnId -> NewClient -> Galley Client
 brigAddClient uid connId client = do
     (brigHost, brigPort) <- brigReq
     r <- call "brig"
         $ method POST
         . host brigHost
         . port brigPort
-        . header "Z-User" (toByteString' uid)
```
Comment (fisx) [inline services/galley/src/Galley/Intra/Client.hs]: oops  :)

File: services/galley/test/integration/API/Teams/LegalHold.hs [hunk 5]
```diff
@@ -260,14 +258,39 @@ testGetLegalHoldDeviceStatus = do
...
-    -- send event to all of user's devices
-    -- send event to all communication peers
-    -- when still pending, notify clients they no longer need approval if deleted when still pending
+    cannon <- view tsCannon
+    WS.bracketR cannon member $ \ws -> withDummyTestServiceForTeam owner tid $ do
+        addTeamMemberInternal tid $ newTeamMember member (rolePermissions RoleMember) Nothing
+        putEnabled tid LegalHoldEnabled
+        requestDevice owner member tid !!! const 204 === statusCode
+        approveLegalHoldDevice member member tid !!! const 200 === statusCode
+        assertExactlyOneLegalHoldDevice member
```
Comment (fisx) [inline services/galley/test/integration/API/Teams/LegalHold.hs]: Also assert LH device number before the operation (same with removal).

File: services/galley/test/integration/API/Teams/LegalHold.hs [hunk 6]
```diff
@@ -585,15 +604,33 @@ approveLegalHoldDevice zusr uid tid = do
...
 
+assertZeroLegalHoldDevices :: HasCallStack  => UserId -> TestM ()
+assertZeroLegalHoldDevices uid = do
+    clients :: [Client] <- getClients uid
+        >>= maybe (error $ "decodeBody: [Client]") pure . decodeBody
+    liftIO $ do
+        let numdevs = length $ clientType <$> clients
+        assertBool ("a legal hold device was found when none was expected for user"
+                    <> show uid)
+                   (numdevs == 0)
```
Comment (fisx) [inline services/galley/test/integration/API/Teams/LegalHold.hs]: You could use `assertEqual` here.

File: services/galley/test/integration/API/Teams/LegalHold.hs [hunk 7]
```diff
@@ -260,14 +258,39 @@ testGetLegalHoldDeviceStatus = do
...
-    pure ()
+testDisableLegalHoldForUser :: TestM ()
+testDisableLegalHoldForUser = do
+    (owner, tid) <- createTeam
+    member <- randomUser
 
-    -- Remove Legal hold device from user.
-    -- send event to all of user's devices
-    -- send event to all communication peers
-    -- when still pending, notify clients they no longer need approval if deleted when still pending
```
Comment (fisx) [inline services/galley/test/integration/API/Teams/LegalHold.hs]: ~~where are these tested? it's ok if they are not, but then they should be changed into TODOs, not removed.~~

ah, found it in the brig tests.  this is why i wanted to have all LH tests here.  can you make one comment in both places cross-referencing the other location of LH tests?

File: services/brig/test/resources/turn/servers-v2.txt (hunk 3 above)
Comment (ChrisPenner) [reply services/brig/test/resources/turn/servers-v2.txt]: Ugh; this happens when you cancel test runs before they complete; it sneaks in for me ALL the time. Thanks

File: services/galley/src/Galley/Intra/Client.hs (hunk 4 above)
Comment (ChrisPenner) [reply services/galley/src/Galley/Intra/Client.hs]: This is intentionaly; we're calling an internal endpoint now, so we don't need the Z-User header. I can leave it if you think we should still have it?

File: services/galley/test/integration/API/Teams/LegalHold.hs (hunk 7 above)
Comment (ChrisPenner) [reply services/galley/test/integration/API/Teams/LegalHold.hs]: I understand the desire to have all tests in one place, but this test is very specific to Brig (adding/removing clients); but yes, I can cross reference somehow :)

File: services/brig/src/Brig/API/Client.hs (hunk 2 above)
Comment (ChrisPenner) [reply services/brig/src/Brig/API/Client.hs]: We have tests which assert this; and in this case it doesn't actually cause any problems (and it's nicer to write), so is it cool if I just leave it?

**Comment by ChrisPenner** [comment]
This also switches `clientId` to `client_id`; referenced in https://github.com/wireapp/wire-ios-sync-engine/pull/1036

File: services/galley/src/Galley/Intra/Client.hs (hunk 4 above)
Comment (fisx) [reply services/galley/src/Galley/Intra/Client.hs]: ah, i misread.  i agree it shouldn't be sent, just thought that this handler was always an internal endpoint and this was there by mistake until now.

anyway, please disregard this thread.  all good!

**Pull request merged**